exchanges/gateio/gateio_wrapper.go

```diff
@@ -2578,3 +2559,215 @@ func (g *Gateio) GetCurrencyTradeURL(_ context.Context, a asset.Item, cp currenc
+			Contract:    s.Pair,
+			Size:        amountWithDirection,
+			Price:       strconv.FormatFloat(s.Price, 'f', -1, 64),
+			ReduceOnly:  s.ReduceOnly,
+			TimeInForce: timeInForce,
+			Text:        s.ClientOrderID,
+		})
+		if err != nil {
+			return nil, err
+		}
+		sr, err := g.deriveFuturesWebsocketOrderResponses(got)
+		if err != nil {
+			return nil, err
+		}
+		return sr[0], nil
+	default:
+		return nil, common.ErrNotYetImplemented
+	}
+}
+
+// deriveSpotWebsocketOrderResponses returns the order submission responses for spot
+func (g *Gateio) deriveSpotWebsocketOrderResponses(responses []WebsocketOrderResponse) ([]*order.SubmitResponse, error) {
+	if len(responses) == 0 {
+		return nil, common.ErrNoResponse
+	}
+
+	out := make([]*order.SubmitResponse, 0, len(responses))
+	for x := range responses {
+		side, err := order.StringToOrderSide(responses[x].Side)
+		if err != nil {
+			return nil, err
+		}
+		status := order.Open
+		if responses[x].FinishAs != "" && responses[x].FinishAs != statusOpen {
+			status, err = order.StringToOrderStatus(responses[x].FinishAs)
+			if err != nil {
+				return nil, err
+			}
+		}
+		oType, err := order.StringToOrderType(responses[x].Type)
+		if err != nil {
+			return nil, err
+		}
+
+		var cost float64
+		var purchased float64
+		if responses[x].AverageDealPrice != 0 {
+			if side.IsLong() {
+				cost = responses[x].FilledTotal.Float64()
+				purchased = responses[x].FilledTotal.Decimal().Div(responses[x].AverageDealPrice.Decimal()).InexactFloat64()
+			} else {
+				cost = responses[x].Amount.Float64()
+				purchased = responses[x].FilledTotal.Float64()
+			}
+		}
+
+		out = append(out, &order.SubmitResponse{
+			Exchange:             g.Name,
+			OrderID:              responses[x].ID,
+			AssetType:            responses[x].Account,
+			Pair:                 responses[x].CurrencyPair,
+			ClientOrderID:        responses[x].Text,
+			Date:                 responses[x].CreateTimeMs.Time(),
+			LastUpdated:          responses[x].UpdateTimeMs.Time(),
+			RemainingAmount:      responses[x].Left.Float64(),
+			Amount:               responses[x].Amount.Float64(),
+			Price:                responses[x].Price.Float64(),
+			AverageExecutedPrice: responses[x].AverageDealPrice.Float64(),
+			Type:                 oType,
+			Side:                 side,
+			Status:               status,
+			ImmediateOrCancel:    responses[x].TimeInForce == iocTIF,
+			FillOrKill:           responses[x].TimeInForce == fokTIF,
+			PostOnly:             responses[x].TimeInForce == pocTIF,
+			Cost:                 cost,
+			Purchased:            purchased,
+			Fee:                  responses[x].Fee.Float64(),
+			FeeAsset:             responses[x].FeeCurrency,
+		})
+	}
+	return out, nil
+}
+
+// deriveFuturesWebsocketOrderResponses returns the order submission responses for futures
+func (g *Gateio) deriveFuturesWebsocketOrderResponses(responses []WebsocketFuturesOrderResponse) ([]*order.SubmitResponse, error) {
```

Comment:
Can we make them pointers? Cos then we can range them on `for _, r := range responses`
```suggestion
func (g *Gateio) deriveFuturesWebsocketOrderResponses(responses []*WebsocketFuturesOrderResponse) ([]*order.SubmitResponse, error) {
```